contracts/examples/erc1155-marketplace/src/lib.rs

```diff
@@ -0,0 +1,386 @@
+		only_owner!(self, "Only owner may call this function!");
+		require!(
+			new_cut_percentage > 0 && new_cut_percentage < PERCENTAGE_TOTAL,
+			"Invalid percentage value, should be between 0 and 100"
+		);
+
+		self.set_percentage_cut(new_cut_percentage);
+
+		Ok(())
+	}
+
+	fn set_token_ownership_contract_address_endpoint(&self, new_address: Address) -> SCResult<()> {
+		only_owner!(self, "Only owner may call this function!");
+		require!(!new_address.is_zero(), "Cannot set to zero address");
+		// TODO: Also check the address with IsSmartContractAddress() once it's added to the API
+
+		self.set_token_ownership_contract_address(&new_address);
+
+		Ok(())
+	}
+
+	// endpoints
+
+	#[payable("EGLD")]
+	#[endpoint]
+	fn bid(
+		&self,
+		type_id: BigUint,
+		token_id: BigUint,
+		#[payment] payment: BigUint,
+	) -> SCResult<()> {
+		require!(
+			self.is_up_for_auction(&type_id, &token_id),
+			"Token is not up for auction"
+		);
+
+		let caller = self.get_caller();
+		let mut auction = self.get_auction_for_token(&type_id, &token_id);
+
+		require!(
+			auction.original_owner != caller,
+			"Can't bid on your own token"
+		);
+		require!(
+			self.get_block_timestamp() < auction.deadline,
+			"Auction ended already"
+		);
+		require!(
+			payment >= auction.min_bid,
+			"Bid must be higher than or equal to the min bid"
+		);
+		require!(
+			payment > auction.current_bid,
+			"Bid must be higher than the current winning bid"
+		);
+		require!(
+			payment <= auction.max_bid,
+			"Bid must be less than or equal to the max bid"
+		);
+
+		// refund losing bid
+		if auction.current_winner != Address::zero() {
+			self.send()
+				.direct_egld(&auction.current_winner, &auction.current_bid, b"bid refund");
+		}
+
+		// update auction bid and winner
+		auction.current_bid = payment;
+		auction.current_winner = caller;
+		self.set_auction_for_token(&type_id, &token_id, &auction);
+
+		Ok(())
+	}
+
+	#[endpoint(endAuction)]
+	fn end_auction(&self, type_id: BigUint, token_id: BigUint) -> SCResult<()> {
+		require!(
+			self.is_up_for_auction(&type_id, &token_id),
+			"Token is not up for auction"
+		);
+
+		let auction = self.get_auction_for_token(&type_id, &token_id);
+
+		require!(
+			self.get_block_timestamp() > auction.deadline || auction.current_bid == auction.max_bid,
+			"auction has not ended yet!"
+		);
+
+		self.clear_auction_for_token(&type_id, &token_id);
+
+		if auction.current_winner != Address::zero() {
+			let percentage_cut = self.get_percentage_cut();
+			let cut_amount = self.calculate_cut_amount(&auction.current_bid, percentage_cut);
+			let amount_to_send = &auction.current_bid - &cut_amount;
+
+			let mut claimable_funds = self.get_claimable_funds();
+			claimable_funds += &cut_amount;
+			self.set_claimable_funds(&claimable_funds);
+
+			// send part of the bid to the original owner
+			self.send().direct_egld(&auction.original_owner, &amount_to_send, b"sold token");
```

Comment:
egld and esdt as well? transfering 2 tokens ?

Reply:
Nope, first is eGLD transfer, then is ERC1155 token transfer